src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,361 @@
+
+ public:
+  expectationNode(){};
+
+  ~expectationNode(){};
+
+  /** A templated function to allow for the creation of an `expectationNode`
+   * instance. The instance created is one with a value and a key. A key,
+   * default value, type, and a flag denoting whether the node is optional are
+   * provided to the underlying constructor. */
+  template <typename T, bool optional = false>
+  expectationNode create(std::string key, T defaultValue) {
+    dataTypeVariant defValVariant = defaultValue;
+    ExpectedType type = static_cast<ExpectedType>(defValVariant.index());
+    // The instance created must have a value so ensure the passed type reflects
+    // requirement
+    if (type == ExpectedType::Valueless) {
+      std::cerr << "[SimEng:ModelConfig] Cannot create a node with a default "
+                   "value and an expected "
+                   "type of `Valueless`"
+                << std::endl;
+      exit(1);
+    }
+    expectationNode node = expectationNode(key, type, optional);
+    node.setDefaultValue(defValVariant);
+    return node;
+  }
+
+  /** A templated function to allow for the creation of an `expectationNode`
+   * instance. The instance created is one with only a key. A key and a flag
+   * denoting whether the node is optional are provided to the underlying
+   * constructor. */
+  template <bool optional = false>
+  expectationNode create(std::string key) {
+    expectationNode node =
+        expectationNode(key, ExpectedType::Valueless, optional);
+    return node;
+  }
+
+  /** A getter function to retrieve the key of a node. */
+  std::string getKey() const { return nodeKey_; }
+
+  /** A getter function to retrieve the held default value of a node. */
+  template <typename T>
+  T getDefault() const {
+    return getByType<T>(defaultValue_);
+  }
+
+  /** A getter function to retrieve the value type of a node. */
+  ExpectedType getType() const { return type_; }
+
+  /** A getter function to retrieve the child expectationNode instances of this
+   * node. */
+  std::vector<expectationNode> getChildren() const { return nodeChildren_; }
+
+  /** A getter function to retrieve whether the expectations should be applied
+   * to a sequence of config values. */
+  bool isSequence() const { return isSequence_; }
+
+  /** A getter function to retrieve whether the node is wild. */
+  bool isWild() const { return isWild_; }
+
+  /** A utility function used by the class to get a value from a `std::variant`
+   * with error handling if the passed type is not currently stored. */
+  template <typename T>
+  T getByType(const dataTypeVariant& variant) const {
+    // Value existence check
+    if (variant.valueless_by_exception()) {
+      std::cerr << "[SimEng:ModelConfig] No value in passed variant within "
+                   "expectationNode with key "
+                << nodeKey_ << std::endl;
+      exit(1);
+    }
+    // Value type check
+    if (!std::holds_alternative<T>(variant)) {
```

Comment:
Also, we may want to use this functionality during runtime which won't trigger the static_assert (I think, just had a cursory glance).

Reply:
I see what you mean, thanks for clarifying. I agree that the runtime check is important. However, I think you should still add the static_assert as it will catch any function call using an **unsupported** type ( any type other than `bool`, `double`, `float`, `int64_t`, `string`, `uint64_t`) during compile time.

Reply:
Okay, I've added the static compile time check.